sentry-core/src/main/java/io/sentry/core/cache/CacheStrategy.java

```diff
@@ -0,0 +1,97 @@
+package io.sentry.core.cache;
+
+import static io.sentry.core.SentryLevel.ERROR;
+
+import io.sentry.core.ISerializer;
+import io.sentry.core.SentryLevel;
+import io.sentry.core.SentryOptions;
+import io.sentry.core.util.Objects;
+import java.io.File;
+import java.nio.charset.Charset;
+import java.util.Arrays;
+import org.jetbrains.annotations.NotNull;
+
+abstract class CacheStrategy {
+
+  @SuppressWarnings("CharsetObjectCanBeUsed")
+  protected static final Charset UTF_8 = Charset.forName("UTF-8");
+
+  protected final @NotNull SentryOptions options;
+  protected final @NotNull ISerializer serializer;
+  protected @NotNull File directory;
+  protected int maxSize;
+
+  CacheStrategy(
+      final @NotNull SentryOptions options,
+      final @NotNull String directoryPath,
+      final int maxSize) {
+    this.options = Objects.requireNonNull(options, "SentryOptions is required.");
+    this.serializer = options.getSerializer();
+
+    Objects.requireNonNull(directoryPath, "Directory is required.");
+    this.directory = new File(directoryPath);
+
+    this.maxSize = maxSize;
+  }
+
+  /**
+   * Check if a Dir. is valid and have write and read permission
+   *
+   * @return true if valid and has permissions or false otherwise
+   */
+  protected boolean isDirectoryValid() {
+    if (!directory.isDirectory() || !directory.canWrite() || !directory.canRead()) {
+      options
+          .getLogger()
+          .log(
+              ERROR,
+              "The directory for caching files is inaccessible.: %s",
+              directory.getAbsolutePath());
+      return false;
+    }
+    return true;
+  }
+
+  /**
+   * Sort files from oldest to the newest using the lastModified method
+   *
+   * @param files the Files
+   */
+  void sortFilesOldestToNewest(@NotNull File[] files) {
+    // just sort it if more than 1 file
+    if (files.length > 1) {
+      // sort by the oldest to the newest
+      Arrays.sort(files, (f1, f2) -> Long.compare(f1.lastModified(), f2.lastModified()));
+    }
+  }
+
+  /**
+   * Rotates the caching folder if full, deleting the oldest files first
+   *
+   * @param files the Files
+   */
+  protected void rotateCacheIfNeeded(final @NotNull File[] files) {
+    final int length = files.length;
+    if (length >= maxSize) {
+      options
+          .getLogger()
+          .log(SentryLevel.WARNING, "Cache folder if full (respecting maxSize). Rotating files");
+      final int totalToBeDeleted = (length - maxSize) + 1;
+
+      sortFilesOldestToNewest(files);
+
+      // delete files from the top of the Array as its sorted by the oldest to the newest
+      for (int i = 0; i < totalToBeDeleted; i++) {
+        final File file = files[i];
+        // sanity check if the file actually exists.
+        if (file.exists()) {
+          if (!file.delete()) {
```

Comment:
M: Since this is inherently racey I ask: Does delete throw here if the file doesn't exist?
If it doesn't, maybe we can skip the `exists` altogether.
If it does, we should wrap this on a try catch anyway.

The `exists` is likely an I/O we could skip since we'll be hitting the disk for the `delete` anyway

Reply:
the problem is to log it, `delete` returns `false` even if the file doesn't exist, it doesn't throw, so we don't know if the file doesn't exist or if it had a problem deleting it.

Reply:
I'd really just get rid of it.

In between checking `exists` and calling `delete` the file could have been deleted so it's anyway not 100%

If delete is false then it didn't delete and there's no much more we can do anyway but to log and go on.